src/generic/stage2_build_tape.h

```diff
@@ -3,515 +3,390 @@
+}
+WARN_UNUSED really_inline bool
+really_parse_number(const uint8_t *const buf, ParsedJson &pj, const uint32_t offset, bool found_minus) {
+  return parse_number(buf, pj, offset, found_minus);
+}
+
+struct structural_parser {
+  const uint8_t* const buf;
+  const size_t len;
+  ParsedJson &pj;
+  uint32_t i; // next structural index
+  uint32_t idx; // location of the structural character in the input (buf)
+  uint8_t c;    // used to track the (structural) character we are looking at
+  uint32_t depth = 0; // could have an arbitrary starting depth
+
+  really_inline structural_parser(const uint8_t *_buf, size_t _len, ParsedJson &_pj, uint32_t _i = 0) : buf{_buf}, len{_len}, pj{_pj}, i{_i} {}
+
+  WARN_UNUSED really_inline int set_error_code(ErrorValues error_code) {
+    pj.error_code = error_code;
+    return error_code;
   }
 
-  UPDATE_CHAR();
-  switch (c) {
-  case '{':
-    pj.containing_scope_offset[depth] = pj.get_current_loc();
-    SET_GOTO_START_CONTINUE();
-    depth++;
-    if (depth >= pj.depth_capacity) {
-      goto fail;
-    }
-    pj.write_tape(
-        0, c); /* strangely, moving this to object_begin slows things down */
-    goto object_begin;
-  case '[':
-    pj.containing_scope_offset[depth] = pj.get_current_loc();
-    SET_GOTO_START_CONTINUE();
-    depth++;
-    if (depth >= pj.depth_capacity) {
-      goto fail;
-    }
-    pj.write_tape(0, c);
-    goto array_begin;
-    /* #define SIMDJSON_ALLOWANYTHINGINROOT
-     * A JSON text is a serialized value.  Note that certain previous
-     * specifications of JSON constrained a JSON text to be an object or an
-     * array.  Implementations that generate only objects or arrays where a
-     * JSON text is called for will be interoperable in the sense that all
-     * implementations will accept these as conforming JSON texts.
-     * https://tools.ietf.org/html/rfc8259
-     * #ifdef SIMDJSON_ALLOWANYTHINGINROOT */
-  case '"': {
-    if (!parse_string(buf, len, pj, depth, idx)) {
-      goto fail;
-    }
-    break;
+  really_inline char advance_char() {
+    idx = pj.structural_indexes[i++];
+    c = buf[idx];
+    return c;
   }
-  case 't': {
-    /* we need to make a copy to make sure that the string is space
-     * terminated.
-     * this only applies to the JSON document made solely of the true value.
-     * this will almost never be called in practice */
+
+  template<typename F>
+  really_inline bool with_space_terminated_copy(const F& f) {
+    /* we need to make a copy to make sure that the string is space terminated.
+      * this is done only for JSON documents made of a sole number
+      * this will almost never be called in practice. We terminate with a space
+      * because we do not want to allow NULLs in the middle of a number (whereas a
+      * space in the middle of a number would be identified in stage 1).
+      *
+      * ED. TODO pretty sure we need to do this in all cases, not just single values,
```

Comment:
Can you sustain this claim? Please see files jsonchecker/fail78.json and 
 jsonchecker/fail79.json that I added and created specifically to answer your comment. If we cannot identify a bug, I suggest to drop the claim. :-)

Reply:
I'll check it in valgrind. A crash should only happen at a page boundary. The basic issue is that parse_atom and parse_number are going to look at the character *past* the last digit, and that doesn't exist in `buf`.

Reply:
It is always safe to read in `buf` up to `len + SIMDJSON_PADDING`. The padding is done at the very start if needed. That's not why a copy is made here. We are making a copy to make sure that we know how the padding was made.

Reply:
Note that in the future, we would like to drop this requirement (that the string has been padded).